pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -139,21 +136,23 @@ func (c azureMSGraphClient) LoginUser(config *v32.AzureADConfig, credential *v32
 	return userPrincipal, groupPrincipals, "", nil
 }
 
-var msGraphRequiredPermissions = []string{
-	"Group.Read.All",
-	"User.Read.All",
-}
-
-// msGraphTokenHasPermissions checks that a token received from Azure AD has the permissions Rancher requires
+// EnsureMSGraphTokenHasPermissions checks that a token received from Azure AD has the permissions Rancher requires
 // for working with the Microsoft Graph API.
-func msGraphTokenHasPermissions(token string) bool {
+func EnsureMSGraphTokenHasPermissions(token string) error {
+	msGraphRequiredPermissions := []string{
+		"Group.Read.All",
+		"User.Read.All",
+	}
+	errMSGraphMissingPermissions := fmt.Errorf("missing required Application type permissions from Microsoft Graph: need %s",
+		strings.Join(msGraphRequiredPermissions, ", "))
+
 	field, err := ExtractFieldFromJWT(token, "roles")
 	if err != nil {
-		return false
+		return errMSGraphMissingPermissions
```

Comment:
`err` here should either be wrapped and returned or logged as info.

Reply:
It is wrapped much later and then logged, since this error falls down all the way to HTTP response sending.

Reply:
I think Kevin is referring to the err produced by the `ExtractField` method. Basically, if we can't extract the roles from the JWT, the user will get an error that they don't have the required permissions, and we will have no idea that the reason why was because our processing of the JWT went wrong.

Reply:
In this case, the processing of the JWT has not gone wrong. This means there are no permissions at all (in this particular case we know it).
If you have only one missing, then the failure will happen later, but the reason the same.
I don't think we need to know which exact permission is missing (although I am open to debate here).
Overall, I don't think any time we return the error from this function we have a JWT processing error.

Reply:
Not sure that I understand properly how we know the jwt processing has not gone wrong. We tried to extract the roles field but got an error - this could be because the token was malformed. How can we be sure that this means the user is missing permissions?

Reply:
Yeah, if it's malformed, then we should indicate so explicitly, will adjust.